### ISSUE TITLE
Sudden `panel.sh` crash, says line194 is the issue

### ISSUE THREAD

**Issue by zenny** [herbstluftwm/herbstluftwm#1595]
Hi @t-wissmann 

Running `panel.sh` crashed in a running session. Tried to rerun it from the command line and got the following error:

`./panel.sh: line 194: /usr/bin/dzen2: cannot execute: required file not found`

The stated line (https://github.com/herbstluftwm/herbstluftwm/blob/c9be0cd1dbc95ff828340af8ebd0f50a6838d5de/share/panel.sh#L194) seems to have pointed anywhere to any file, I guess.

Thanks in advance,

Cheers,
zenny

PS: I love the simplicity of panel.sh (tried with polybar and got distracted with icons and colors due to my personal eye issues, called tetrachromat). So panel.sh serves well for those who has accessibility issues.

**Comment by t-wissmann** [comment]
It seems that `dzen2` is not installed (anymore) on your system. What is the output of these commands?
```bash
which dzen2
```
and
```bash
dzen2 -help
```

Also, have you considered switching to [barpyrus](https://github.com/t-wissmann/barpyrus)? It is  more robust because it's written in python and it also allows a system tray :-)

**Comment by zenny** [comment]
> It seems that `dzen2` is not installed (anymore) on your system. What is the output of these commands?
> 
> ```shell
> which dzen2
> ```
> 

dzen2 is installed as of follow:

```
$ which dzen2
/usr/bin/dzen2
```

voidlinux-specific query:

```
$ xbps-query -v -Rs dzen2
[*] dzen2-20130923_2 X notification utility
```


> and
> 
> ```shell
> dzen2 -help
> ```
The output in the shell for help is:

```
$ dzen2 -help
zsh: command not found: dzen2
```

So I force-installed `dzen2` again:

```
# xbps-install -Sf dzen2
[*] Updating repository `https://repo-default.voidlinux.org/current/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/multilib/nonfree/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/multilib/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/nonfree/x86_64-repodata' ...

Name  Action    Version           New version            Download size
dzen2 reinstall 20130923_2        20130923_2             -

Size required on disk:         120KB
Space available on disk:        11GB

Do you want to continue? [Y/n] y

[*] Downloading packages
dzen2-20130923_2.x86_64.xbps.sig2: 512B [avg rate: 20MB/s]
dzen2-20130923_2.x86_64.xbps: 29KB [avg rate: 465MB/s]
dzen2-20130923_2: verifying RSA signature...

[*] Collecting package files
dzen2-20130923_2: collecting files...

[*] Unpacking packages
dzen2-20130923_2: unpacking ...

[*] Configuring unpacked packages
dzen2-20130923_2: configuring ...
dzen2-20130923_2: installed successfully.

0 downloaded, 1 installed, 0 updated, 1 configured, 0 removed.
```

Yet, it cannot launch panel:

```
$ ~/.config/herbstluftwm/panel.sh
/home/zenny/.config/herbstluftwm/panel.sh: line 194: /usr/bin/dzen2: cannot execute: required file not found
```

> 
> Also, have you considered switching to [barpyrus](https://github.com/t-wissmann/barpyrus)? It is more robust because it's written in python and it also allows a system tray :-)

@t-wissmann, Thank you so much for the pointer, but three peculiar accessibility issues [tetrachromat (https://www.youtube.com/watch?v=_OUiOpw1UC4), palinopsia or afterimage (https://www.youtube.com/watch?v=Va4KhD3WemQ) and photophobia (light-sensitivity)] that I have prevents me from choosing extremely colorful interfaces. Cases like mine prefer a simple `panel.sh`, fyi.

**Comment by zenny** [comment]
> > It seems that `dzen2` is not installed (anymore) on your system. What is the output of these commands?
> > ```shell
> > which dzen2
> > ```
> 
> dzen2 is installed as of follow:
> 
> ```
> $ which dzen2
> /usr/bin/dzen2
> ```
> 
> voidlinux-specific query:
> 
> ```
> $ xbps-query -v -Rs dzen2
> [*] dzen2-20130923_2 X notification utility
> ```
> 
> > and
> > ```shell
> > dzen2 -help
> > ```
> 
> The output in the shell for help is:
> 
> ```
> $ dzen2 -help
> zsh: command not found: dzen2
> ```
> 
> So I force-installed `dzen2` again:
> 
> ```
> # xbps-install -Sf dzen2
> [*] Updating repository `https://repo-default.voidlinux.org/current/x86_64-repodata' ...
> [*] Updating repository `https://repo-default.voidlinux.org/current/multilib/nonfree/x86_64-repodata' ...
> [*] Updating repository `https://repo-default.voidlinux.org/current/multilib/x86_64-repodata' ...
> [*] Updating repository `https://repo-default.voidlinux.org/current/nonfree/x86_64-repodata' ...
> 
> Name  Action    Version           New version            Download size
> dzen2 reinstall 20130923_2        20130923_2             -
> 
> Size required on disk:         120KB
> Space available on disk:        11GB
> 
> Do you want to continue? [Y/n] y
> 
> [*] Downloading packages
> dzen2-20130923_2.x86_64.xbps.sig2: 512B [avg rate: 20MB/s]
> dzen2-20130923_2.x86_64.xbps: 29KB [avg rate: 465MB/s]
> dzen2-20130923_2: verifying RSA signature...
> 
> [*] Collecting package files
> dzen2-20130923_2: collecting files...
> 
> [*] Unpacking packages
> dzen2-20130923_2: unpacking ...
> 
> [*] Configuring unpacked packages
> dzen2-20130923_2: configuring ...
> dzen2-20130923_2: installed successfully.
> 
> 0 downloaded, 1 installed, 0 updated, 1 configured, 0 removed.
> ```
> 
> Yet, it cannot launch panel:
> 
> ```
> $ ~/.config/herbstluftwm/panel.sh
> /home/zenny/.config/herbstluftwm/panel.sh: line 194: /usr/bin/dzen2: cannot execute: required file not found
> ```
> 
> > Also, have you considered switching to [barpyrus](https://github.com/t-wissmann/barpyrus)? It is more robust because it's written in python and it also allows a system tray :-)
> 
> @t-wissmann, Thank you so much for the pointer, but three peculiar accessibility issues [tetrachromat (https://www.youtube.com/watch?v=_OUiOpw1UC4), palinopsia or afterimage (https://www.youtube.com/watch?v=Va4KhD3WemQ) and photophobia (light-sensitivity)] that I have prevents me from choosing extremely colorful interfaces. Cases like mine prefer a simple `panel.sh`, fyi.

@t-wissmann UPDATE: Btw, I tried earlier `barpyrus` in default mode, but what it lacks is the display of the window information on top as you can see in the `panel.py` and it gave a series of `Invalid font index 3` errors, fyi.

**Comment by t-wissmann** [comment]
It's strange: `which dzen2` reports that the `dzen2` executable exists but then actually executing it reports "file not found". Maybe `/usr/bin/dzen2` is a broken link? (Just guessing)... what is the output of this command?
```bash
/bin/ls -alt `which dzen2`
```

Regarding the window title: there is a `HLWMWindowTitle` widget in barpyrus: https://github.com/t-wissmann/barpyrus/blob/master/barpyrus/hlwm.py#L248C7-L248C22 and it's used in `share/config.py`: https://github.com/t-wissmann/barpyrus/blob/master/share/config.py

**Comment by zenny** [comment]
> It's strange: `which dzen2` reports that the `dzen2` executable exists but then actually executing it reports "file not found". Maybe `/usr/bin/dzen2` is a broken link? (Just guessing)... what is the output of this command?
> 
> ```shell
> /bin/ls -alt `which dzen2`
> ```
> 

```
$ /bin/ls -alt `which dzen2`
-rwxr-xr-x 1 root root 56320 Nov 25  2018 /usr/bin/dzen2
```

> Regarding the window title: there is a `HLWMWindowTitle` widget in barpyrus: https://github.com/t-wissmann/barpyrus/blob/master/barpyrus/hlwm.py#L248C7-L248C22 and it's used in `share/config.py`: https://github.com/t-wissmann/barpyrus/blob/master/share/config.py

Thanks for the pointer, I shall have a look.

**Comment by t-wissmann** [comment]
Can you list the contents of the installed dzen2 package? It seems as if `/usr/bin/dzen2` was not updated: you just reinstalled the package but  `/usr/bin/dzen2` was last changed in 2018. What is the output of this command?
```bash
ldd /usr/bin/dzen2
```
It checks whether all libraries required by `dzen2` are present. But all in all, it is an issue with dzen2 and/or the package. Panel.sh is suffering from it but it it's not an issue I can solve in `panel.sh`.

**Comment by The-Compiler** [comment]
Note that bash would fail with `bash: /bin/doesnotexist: No such file or directory` if the binary wasn't there. I've never seen `bash: /bin/doesnotexist: No such file or directory`, but the issue is something different:

- [Linux Bash Shell Script Error: cannot execute: required file not found - Unix & Linux Stack Exchange](https://unix.stackexchange.com/questions/721844/linux-bash-shell-script-error-cannot-execute-required-file-not-found)
- [bash - Cannot exceute binary: required file not found - Unix & Linux Stack Exchange](https://unix.stackexchange.com/questions/753458/cannot-exceute-binary-required-file-not-found)

Given that you can't run `dzen2` in a terminal either, this is definitely something about your setup and not hlwm related.

Next steps I'd take to find out more:

- `file $(which dzen2)` -> would expect something along the lines of "ELF 64-bit LSB pie executable"
- If it's indeed that, `ldd $(which dzen2)`
- If it's a wrapper script of some sorts, `head $(which dzen2)` and `head $(which dzen2) | hexdump -C`

**Comment by zenny** [comment]
> Can you list the contents of the installed dzen2 package? It seems as if `/usr/bin/dzen2` was not updated: you just reinstalled the package but `/usr/bin/dzen2` was last changed in 2018.

dzen2 has not been updated since 2015. And the last void package was released in 2018 as seen below.

```
$ xbps-query -v -R dzen2
architecture: x86_64
build-date: 2018-11-25 03:35 CET
filename-sha256: 7515bd005296dd225e6213c6a358e50b9a760cdeb218b9b11b9c41efd63657a6
filename-size: 29KB
homepage: https://github.com/robm/dzen
installed_size: 120KB
license: MIT
maintainer: Juan RP <xtraeme@voidlinux.eu>
pkgname: dzen2
pkgver: dzen2-20130923_2
repository: https://repo-default.voidlinux.org/current
reverts:
        20140623_1
run_depends:
        libXpm>=3.5.7_1
        libX11>=1.2_1
        libXinerama>=1.0.3_1
        libXft>=2.1.13_1
        glibc>=2.28_1
shlib-requires:
        libXpm.so.4
        libX11.so.6
        libXinerama.so.1
        libXft.so.2
        libc.so.6
short_desc: X notification utility
source-revisions: dzen2:3b9cd990c9
```
> What is the output of this command?
> 
> ```shell
> ldd /usr/bin/dzen2
> ```

```
$ ldd /usr/bin/dzen2
        linux-vdso.so.1 (0x00007ffd849fe000)
        libXpm.so.4 => /usr/lib/libXpm.so.4 (0x00007fefb2c8e000)
        libX11.so.6 => /usr/lib/libX11.so.6 (0x00007fefb2b49000)
        libXinerama.so.1 => /usr/lib/libXinerama.so.1 (0x00007fefb2b44000)
        libXft.so.2 => /usr/lib/libXft.so.2 (0x00007fefb2b2a000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007fefb2940000)
        libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007fefb2916000)
        libXext.so.6 => /usr/lib/libXext.so.6 (0x00007fefb28ff000)
        libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x00007fefb28b4000)
        libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x00007fefb27f9000)
        libXrender.so.1 => /usr/lib/libXrender.so.1 (0x00007fefb27ec000)
        /lib/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fefb2d6a000)
        libXau.so.6 => /usr/lib/libXau.so.6 (0x00007fefb27e7000)
        libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007fefb27df000)
        libexpat.so.1 => /usr/lib/libexpat.so.1 (0x00007fefb27b2000)
        libz.so.1 => /usr/lib/libz.so.1 (0x00007fefb2798000)
        libbz2.so.1 => /usr/lib/libbz2.so.1 (0x00007fefb2785000)
        libpng16.so.16 => /usr/lib/libpng16.so.16 (0x00007fefb274f000)
        libbrotlidec.so.1 => /usr/lib/libbrotlidec.so.1 (0x00007fefb2741000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007fefb265d000)
        libbrotlicommon.so.1 => /usr/lib/libbrotlicommon.so.1 (0x00007fefb263a000)
```
> 
> It checks whether all libraries required by `dzen2` are present. But all in all, it is an issue with dzen2 and/or the package. Panel.sh is suffering from it but it it's not an issue I can solve in `panel.sh`.

Thanks for your input.

**Comment by t-wissmann** [comment]
The `ldd` output looks fine, too. I have no idea why `dzen2 -help` fails; please also see the suggestions by @The-Compiler above :-)

**Comment by zenny** [comment]
> Note that bash would fail with `bash: /bin/doesnotexist: No such file or directory` if the binary wasn't there. I've never seen `cannot execute: required file not found`, but the issue is something different:
> 
>     * [Linux Bash Shell Script Error: cannot execute: required file not found - Unix & Linux Stack Exchange](https://unix.stackexchange.com/questions/721844/linux-bash-shell-script-error-cannot-execute-required-file-not-found)
> 
>     * [bash - Cannot exceute binary: required file not found - Unix & Linux Stack Exchange](https://unix.stackexchange.com/questions/753458/cannot-exceute-binary-required-file-not-found)
> 
> 
> Given that you can't run `dzen2` in a terminal either, this is definitely something about your setup and not hlwm related.

Honestly speaking, the panel was running any issue for years, but it crashed all of a sudden as I explained in OT.

> 
> Next steps I'd take to find out more:
> 
>     * `file $(which dzen2)` -> would expect something along the lines of "ELF 64-bit LSB pie executable"

```
$ file $(which dzen2)
/usr/bin/dzen2: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=8150eb6cc14a73947c8219313930729eab4e99e6, stripped
```

> 
>     * If it's indeed that, `ldd $(which dzen2)` and 

```
$ ldd $(which dzen2)
        linux-vdso.so.1 (0x00007fff4a195000)
        libXpm.so.4 => /usr/lib/libXpm.so.4 (0x00007f9711c63000)
        libX11.so.6 => /usr/lib/libX11.so.6 (0x00007f9711b1e000)
        libXinerama.so.1 => /usr/lib/libXinerama.so.1 (0x00007f9711b19000)
        libXft.so.2 => /usr/lib/libXft.so.2 (0x00007f9711aff000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f9711915000)
        libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007f97118eb000)
        libXext.so.6 => /usr/lib/libXext.so.6 (0x00007f97118d4000)
        libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x00007f9711889000)
        libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x00007f97117ce000)
        libXrender.so.1 => /usr/lib/libXrender.so.1 (0x00007f97117c1000)
        /lib/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f9711d3f000)
        libXau.so.6 => /usr/lib/libXau.so.6 (0x00007f97117bc000)
        libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007f97117b4000)
        libexpat.so.1 => /usr/lib/libexpat.so.1 (0x00007f9711787000)
        libz.so.1 => /usr/lib/libz.so.1 (0x00007f971176d000)
        libbz2.so.1 => /usr/lib/libbz2.so.1 (0x00007f971175a000)
        libpng16.so.16 => /usr/lib/libpng16.so.16 (0x00007f9711724000)
        libbrotlidec.so.1 => /usr/lib/libbrotlidec.so.1 (0x00007f9711716000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f9711632000)
        libbrotlicommon.so.1 => /usr/lib/libbrotlicommon.so.1 (0x00007f971160f000)
```

> `LD_DEBUG=all dzen2 --help`

```
$ LD_DEBUG=all dzen2 --help
zsh: command not found: dzen2
$ bash
$ LD_DEBUG=all dzen2 --help
bash: /usr/bin/dzen2: cannot execute: required file not found
```
> 
>     * If it's a wrapper script of some sorts, `head $(which dzen2)` 

```
$ head $(which dzen2)
ELF>U@@8
        @@@h))000oo     (((0DDPtd44QtdRtd/lib/ld-linux-x86-64.so.2GNUGNUPlJs|190rNrA
 vem|BE)qX9[{6@5 =){zK~N;m\)A
 V ;m^NU4tPDJ.k -(Xp{>e^rx'"V  libXpm.so.4_ITM_deregisterTMCloneTableXFreeColors__gmon_start__XCreateGCXFreeGC_ITM_registerTMCloneTableXCreatePixmapXpmReadFileToPixmapXpmFreeAttributeslibX11.so.6XFillArcXGetWindowAttributesXSetForegroundXGrabKeyboardXFreeStringListXDrawArcXNextEventXGrabPointerXUnmapWindowXExtentsOfFontSetXDestroyWindowXSyncXMapWindowXStringToKeysymXLoadQueryFontXFreeFontXUngrabKeyboardXCreateFontSetXCreateWindowXMoveResizeWindowXLowerWindowXrmGetResourceXAllocNamedColorXFreeXDrawRectangleXResizeWindowXSetClassHintXInternAtomXChangePropertyXLookupStringXRaiseWindowXStringListToTextPropertyXPendingXResourceManagerStringXCopyAreaXSupportsLocaleXrmInitializeXrmDestroyDatabaseXFontsOfFontSetXFreePixmapXSetBackgroundXUngrabPointerXCopyPlaneXrmGetStringDatabaseXParseGeometryXStoreNameXOpenDisplayXReadBitmapFileXFillRectangleXCloseDisplayXMapRaisedXSetWMClientMachineXAllocClassHintlibXinerama.so.1XineramaIsActiveXineramaQueryScreenslibXft.so.2XftColorAllocNameXftFontCloseXftDrawDestroyXftTextExtentsUtf8XftFontOpenNameXftDrawCreateXftFontOpenXlfdXftDrawStringUtf8libc.so.6fflush__printf_chkexitexeclsetlocalestrncmpwaitperror__isoc99_sscanf__fdelt_chkstrncpyputsfork__stack_chk_failselectgetpidstrdupstrtokstrtolstrlensetitimersigemptysetstrstrread__fprintf_chkstdoutstrtok_rmemcpystrtoulmallocstrcat__strncpy_chkgetenvstderrgethostnamefwritesigactionstrchr__
6tirintf@uik__cxpPVinalizesetsidstrcmp__libc_start_mainfree_edata__bss_start_endGLIBC_2.15GLIBC_2.14GLIBC_2.4GLIBC_2.7GLIBC_2.3.4GLIBC_2.2.5n!ii
'/7?{^V
     (.rwupx

 (08HPX `!h"p#x$%&')*+,-/012345689: ;(<0=8>@?H@PAXB`ChDpExFGHIJKLMNOPQRSTUVXYZ [(\0]8^@_H`PaXb`chdpexfghijklmnopqHHt5:%<@%:h%2h%*h%"h%h%h%
h%hp%`%h        P%h
@%h
   0%ګh
%ʫh%«␤%h%h%h%h%h%h%h%hp%zh`%rhP%jh@%bh%Zh %Rh%Jh%Bh%:h %2h!%*h"%"h#%h$%h%%
h&%h'p%h(`%h)P%h*@%h+0%ڪh, %Ҫh-%ʪh.%ªh/%h0%h1%h2%h3%h4%h5%h6%h7p%zh8`%rh9P%jh:@%bh;0%Zh< %Rh=%Jh>%Bh?%:h@%2hA%*hB%"hC%hD%hE%
hF%hGp%hH`%hIP%hJ@%hK0%کhL %ҩhM%ʩhN%©hO%hP%hQ%hR%hS%hT%hU%hV%hWp%zhX`%rhYP%jhZ@%bh[0%Zh\ %Rh]%Jh^%Bh_%:h`%2ha%*hb%"hc%hd%he%
hf%hgp%hh`%hiP%hj@%hk0%ڨhl %fAWA1AVIAUATUSH8dH%(H$(1Hj&HHjH1HiHHjHHjH&HOcHHsjH
                                                                              H5IH=HHL$LH*jH5%jHH$IHH$WL$LHHiH5iL$LHHiH5i]L$LHHiH5i7L$LHH}iH5iHHD$H-iL%iH$L-iDAfA9A9HcӹHML
                                                                                                                                                                          Ltù
LLA9~HcHT$pH$I<Ht$hL$
                     t
```

> and `head $(which dzen2) | hexdump -C`

The output is posted in https://termbin.com/07u1

Thanks @The-Compiler.

**Comment by zenny** [comment]
> Regarding the window title: there is a `HLWMWindowTitle` widget in barpyrus: https://github.com/t-wissmann/barpyrus/blob/master/barpyrus/hlwm.py#L248C7-L248C22 and it's used in `share/config.py`: https://github.com/t-wissmann/barpyrus/blob/master/share/config.py

@t-wissmann is there any chance that I can change to a more legible bitmap-font that looks like Inconsolata or Roboto or FiraCode in `barpyrus`? I did see a few lines in the `lemonbar.py`, but could not exactly figure out how? Appreciated.

**Comment by zenny** [comment]
Instead of using voidlinux xbps-binary of `dzen2`, I compiled manually by pulling from https://github.com/robm/dzen repo and installed in `/usr/local/bin` and `dzen1 --help` worked and `panel.sh` got launched without a problem. 

Two things that went above my head:

1. The same xbps-binary of `dzen2` from voidlinux upstream has been working for years and stopped working without any updates?!
2. Manually compiling the same source as of the voidlinux upstream pointer works flawlessly?!

Something seems to be breaking in voidlinux upstream it seems which was also in the case of `emacs` as I reported in https://github.com/void-linux/void-packages/issues/51727.

Anyway thanks to @The-Compiler and @t-wissmann.
